### DISCUSSION TITLE
Idea on how to solve an ambiguous alternative

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1676] [Q&A] [answered]
Hello,

I converted my grammar from Xtext to Langium and when I start the VSCode extension I have this message:
```
Ambiguous Alternatives Detected: <0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1> in <OR1> inside <Operation​> Rule,
<@, ID, @, ID, ID, ID, (, ID> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
```

The error comes from the Operation rule defined [here](https://github.com/ydaveluy/xsmp-modeler/blob/19968aa7b5d1e181a0470693364f4e84bc9cf863/src/language/xsmpcat.langium#L450).

The parser works as expected but I would like to fix the ambiguity.
I have tried many approaches and either the parser is not parsing the code correctly or there is an ambiguity in the grammar.

I wrote a simplified grammar in the [playground](https://langium.org/playground?grammar=OYJwhgthYgBAEgUwDbIPYHU0mQEwFD6IB2ALiAJ6wCyauKAXPrC7ABSkUAOiAzgNQBeACrdEsAD6w0PcKQCWaYkIDyssAqUBKAFQBuQqJ4NYABRDyI8hQDdERxBICCpcvIBGAV1L2xB85bW8nYOJgDkXBZWtohhsMSQiIIAkgAiBi5uXj6hsGEaWd6x8Ykp6YSZFtmITKx5AAJhnDyCANqVHkWhaQC6sPgG%2BGqIcorEsLWsBVVFAoId1TrMrGH0AGZxbCCIpJ4gxKYwiT4ggmwASjt7B0cQOyMSl7v7h%2BB3JxjWABYAcolaLCkYRsaHkuDCAISdzKeTYcQhYQGhCe11exxGkxY006PjmCyKOhYyxYzSSrQC0WCvmMvSR%2BBRL1u9xAn1Iv0SmNg2OqeNcMx8hP6dVJbQpQRCYgYvRK0LSdPwXzB9HGJysCWQsAwAGUTAB6AA6Al1BlV8nVsDSetaAH0wABaABeTjtAC0eq19QB3a09HTGwiK3DK2Cm83UAAy1oAwipqNQAKI-YR6-UGnQe3j6rW%2BgD8%2Bp0qf9geDobAGq1kZjccTydgBvrrQAevriPqQL7-UA&content=A4JwlgtmAuYG4FMAEAhA9mgNgKAIbWnACMBXaZAYTQDsBnaPA4s5FATwAU0xryRtsAekFJc1GmwhoStJCATQSIakmC4QuCAoT8AAlToMAJggBmSXey48%2BqDJiRpgO-GBoA5TQgAUASgHCcgpKKmoaWrYA7jAAFkjUXtj6NPTYJuaWnNy8OnZYQYrKHOpefJ5ajs4asB5efgEicNxGlS411GlmSE1gLT1GAPJVrjR%2BQA).

Any idea on how to update the grammar (I cannot change the output format that is already used in production) ?

**Comment by msujew** [comment, marked as answer]
Hey @ydaveluy,

I'm not sure it's possible to resolve this ambiguity error with the tools provided by Langium. The parser works right now, mostly by abusing the ambiguity behavior (i.e. picking the first option in case of a detected ambiguity). In theory, you would need a manual lookahead that looks a few tokens into the token stream to detect whether there's another `ID` token right after the current `ID` token. However, Langium doesn't provide the API for it.

**Comment by ydaveluy** [comment]
Ok thanks, I will keep it as is as the parser works properly.
